File: engine/python/conform.py
#!/usr/bin/env python3
"""Python conformance entrypoint (SPEC/20-b1-canon-1.md §5).

    stdin  : a JSON document
    stdout : 64 lowercase hex digits + newline
    stderr : a B1_ERR_* token when the document is rejected
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from b1_quality.canon import B1Error, digest_text  # noqa: E402


def main() -> int:
    text = sys.stdin.read()
    try:
        sys.stdout.write(digest_text(text) + "\n")
        return 0
    except B1Error as err:
        sys.stderr.write(err.token + "\n")
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: surfaces/web/conform.mjs
// JavaScript conformance entrypoint (SPEC/20-b1-canon-1.md §5).
//
//   stdin  : a JSON document
//   stdout : 64 lowercase hex digits + newline
//   stderr : a B1_ERR_* token when the document is rejected
//
// Uses the same canon.mjs the browser dashboard loads, so what conformance verifies is the code the
// review surface actually runs — not a second implementation that happens to sit beside it.

import { readFileSync } from "node:fs";
import { digestText, B1Error } from "./canon.mjs";

const raw = readFileSync(0);

// Strict UTF-8: the default decode substitutes U+FFFD, which would silently change the document
// and therefore its digest.
let text;
try {
  text = new TextDecoder("utf-8", { fatal: true }).decode(raw);
} catch {
  process.stderr.write("B1_ERR_INVALID_UTF8\n");
  process.exit(2);
}

try {
  process.stdout.write(digestText(text) + "\n");
} catch (err) {
  process.stderr.write((err instanceof B1Error ? err.token : "B1_ERR_PARSE") + "\n");
  process.exit(2);
}
